pkg/cvo/upgradeable.go

```diff
@@ -0,0 +1,215 @@
+			ret = append(ret, c)
+		}
+	}
+	sort.Slice(ret, func(i, j int) bool { return ret[i].Type < ret[j].Type })
+	return ret
+}
+
+// setUpgradeable updates the currently calculated status of Upgradeable
+func (optr *Operator) setUpgradeable(u *upgradeable) {
+	if u != nil {
+		u.At = time.Now()
+	}
+
+	optr.upgradeableStatusLock.Lock()
+	defer optr.upgradeableStatusLock.Unlock()
+	optr.upgradeable = u
+}
+
+// getUpgradeable returns the current calculated status of upgradeable. It
+// may be nil.
+func (optr *Operator) getUpgradeable() *upgradeable {
+	optr.upgradeableStatusLock.Lock()
+	defer optr.upgradeableStatusLock.Unlock()
+	return optr.upgradeable
+}
+
+type upgradeableCheck interface {
+	// returns a not-nil condition when the check fails.
+	Check() *configv1.ClusterOperatorStatusCondition
+}
+
+type clusterOperatorsUpgradeable struct {
+	coLister configlistersv1.ClusterOperatorLister
+}
+
+func (check *clusterOperatorsUpgradeable) Check() *configv1.ClusterOperatorStatusCondition {
+	cond := &configv1.ClusterOperatorStatusCondition{
+		Type:   configv1.ClusterStatusConditionType("UpgradeableClusterOperators"),
+		Status: configv1.ConditionFalse,
+	}
+	ops, err := check.coLister.List(labels.Everything())
+	if meta.IsNoMatchError(err) {
+		return nil
+	}
+	if err != nil {
+		cond.Reason = "FailedToListClusterOperators"
+		cond.Message = errors.Wrap(err, "failed to list cluster operators").Error()
+		return cond
+	}
+
+	var notup []string
+	for _, op := range ops {
+		if up := resourcemerge.FindOperatorStatusCondition(op.Status.Conditions, configv1.OperatorUpgradeable); up != nil && up.Status == configv1.ConditionFalse {
+			notup = append(notup, fmt.Sprintf("Cluster Operator %q is not Upgradeable(%q): %s", op.GetName(), up.Reason, up.Message))
+		}
+	}
+
+	if len(notup) == 0 {
+		return nil
+	}
+	msg := ""
+	reason := ""
+	if len(notup) == 1 {
+		msg = notup[0]
+		reason = "ClusterOperatorNotUpgradeable"
+	} else {
+		msg = fmt.Sprintf("Multiple Cluster Operators are not upgradeable:\n* %s", strings.Join(notup, "\n* "))
+		reason = "ClusterOperatorsNotUpgradeable"
+	}
+	cond.Reason = reason
+	cond.Message = msg
+	return cond
+}
+
+type clusterVersionOverridesUpgradebale struct {
+	name     string
+	cvLister configlistersv1.ClusterVersionLister
+}
+
+func (check *clusterVersionOverridesUpgradebale) Check() *configv1.ClusterOperatorStatusCondition {
+	cond := &configv1.ClusterOperatorStatusCondition{
+		Type:   configv1.ClusterStatusConditionType("UpgradeableClusterVersionOverrides"),
+		Status: configv1.ConditionFalse,
+	}
+
+	cv, err := check.cvLister.Get(check.name)
+	if meta.IsNoMatchError(err) || apierrors.IsNotFound(err) {
+		return nil
+	}
+
+	overrides := false
+	for _, o := range cv.Spec.Overrides {
+		if o.Unmanaged {
+			overrides = true
+		}
+	}
+	if !overrides {
+		return nil
+	}
+
+	cond.Reason = "ClusterVersionOverridesSet"
+	cond.Message = "Cluster Version overrides is being used to unmanage certain owned resources."
```

Comment:
Message should probably be `Disabling ownership via cluster version overrides prevents upgrades.  Please remove overrides before continuing.`

Reply:
+1 fixed to match your suggestion.